pkg/providers/bedrock/chat.go

```diff
@@ -0,0 +1,140 @@
+package bedrock
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	//"glide/pkg/providers/clients"
+
+	"glide/pkg/api/schemas"
+
+	"go.uber.org/zap"
+
+	"github.com/google/uuid"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime"
+)
+
+type ChatMessage struct {
+	Role    string `json:"role"`
+	Content string `json:"content"`
+}
+
+// ChatRequest is an Bedrock-specific request schema
+type ChatRequest struct {
+	Messages             string               `json:"inputText"`
+	TextGenerationConfig TextGenerationConfig `json:"textGenerationConfig"`
+}
+
+type TextGenerationConfig struct {
+	Temperature   float64  `json:"temperature"`
+	TopP          float64  `json:"topP"`
+	MaxTokenCount int      `json:"maxTokenCount"`
+	StopSequences []string `json:"stopSequences,omitempty"`
+}
+
+// NewChatRequestFromConfig fills the struct from the config. Not using reflection because of performance penalty it gives
+func NewChatRequestFromConfig(cfg *Config) *ChatRequest {
+	return &ChatRequest{
+		TextGenerationConfig: TextGenerationConfig{
+			MaxTokenCount: cfg.DefaultParams.MaxTokens,
+			StopSequences: cfg.DefaultParams.StopSequence,
+			Temperature:   cfg.DefaultParams.Temperature,
+			TopP:          cfg.DefaultParams.TopP,
+		},
+	}
+}
+
+func NewChatMessagesFromUnifiedRequest(request *schemas.UnifiedChatRequest) string {
+	// message history not yet supported for AWS models
+	message := fmt.Sprintf("Role: %s, Content: %s", request.Message.Role, request.Message.Content)
+
+	return message
+}
+
+// Chat sends a chat request to the specified bedrock model.
+func (c *Client) Chat(ctx context.Context, request *schemas.UnifiedChatRequest) (*schemas.UnifiedChatResponse, error) {
+	// Create a new chat request
+	chatRequest := c.createChatRequestSchema(request)
+
+	chatResponse, err := c.doChatRequest(ctx, chatRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(chatResponse.ModelResponse.Message.Content) == 0 {
+		return nil, ErrEmptyResponse
+	}
+
+	return chatResponse, nil
+}
+
+func (c *Client) createChatRequestSchema(request *schemas.UnifiedChatRequest) *ChatRequest {
+	// TODO: consider using objectpool to optimize memory allocation
+	chatRequest := c.chatRequestTemplate // hoping to get a copy of the template
+	chatRequest.Messages = NewChatMessagesFromUnifiedRequest(request)
+
+	return chatRequest
+}
+
+func (c *Client) doChatRequest(ctx context.Context, payload *ChatRequest) (*schemas.UnifiedChatResponse, error) {
+	rawPayload, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal chat request payload: %w", err)
+	}
+
+	cfg, _ := config.LoadDefaultConfig(ctx,
```

Comment:
We don't want this authN to happen every time a new request comes. I would rather authorize the bedrock client once and store as a property of Glide's client struct.

Reply:
How can we do that?

Reply:
Just need to move the `client := bedrockruntime.NewFromConfig(cfg)` line (and related code it needs) to https://github.com/EinStack/glide/pull/141/files/89748889757c3cffbda83c44bbbc020b12494af5#diff-35d53e40a56aa1f71c3074a67f8c671c595cca1dae702e77b0ed6f29911066ebR38-R55 so bedrock_client will sit there and we will call from there in this method. Let me know if that makes sense.